regularizer_new.py

```diff
@@ -103,83 +245,249 @@ def tokenizer_helper(s, add_special_tokens=True):
+                    else:
+                        # (+d(...))
+                        end_brack = 2
+
+                    tot_score = 0
+                    for k in range(st+2, en-1):
+                        cand_score = score_chart[(st+2,k)] + score_chart[(k+1,en-1)]
+                        tot_score += cand_score
+
+                    s1 = recurse(score_chart, curr_str, st+2, st+end_brack)
+                    s2 = recurse(score_chart, curr_str, st + end_brack + 1, en-1)
+                    
+                    if mean:
+                        norm_score = chart[(st+2, st+end_brack)] + chart[(st + end_brack + 1, en-1)] - (tot_score/(en-st-3))
+                    else:
+                        k_rand = randint(st+2, en-3)
+                        norm_score = chart[(st+2, st+end_brack)] + chart[(st + end_brack + 1, en-1)] - score_chart[(st+2,k_rand)] - score_chart[(k_rand+1,en-2)]
                 else:
-                    k_rand = randint(st, en-1)
-                    norm_score = best_score - score_chart[(st,k_rand)] - score_chart[(k_rand+1,en)]
-        
+                    # Normal computation
+                    scores = []
+                    for k in range(st, en):
+                        cand_score = score_chart[(st,k)] + score_chart[(k+1,en)]
+                        scores.append(cand_score)
+                    
+                    scores = torch.tensor(scores, requires_grad = True, dtype = torch.float).to(device)
+                    if (self.gumbel):
+                        # Soft split decision
+                        softmaxed = F.gumbel_softmax(scores, tau=tau, hard=True)
+                        best = (torch.argmax(softmaxed) + st).item()
+                        best_score = torch.sum(softmaxed*scores)
+                        tot_score = torch.sum(scores)
+                    else:
+                        # Greedy split decision
+                        best = (torch.argmax(scores) + st).item()
+                        best_score = torch.max(scores)
+                        tot_score = torch.sum(scores)
+
+                    if (print_parse):
+                        if (self.spaces):
+                            word_list = curr_str.split(" ")
+                            print((" ").join(word_list[st:best+1]), (" ").join(word_list[best+1:en]))
+                        else:
+                            print(curr_str[st:best+1], curr_str[best+1:en])
+
+                    s1 = recurse(score_chart, curr_str, st, best)
+                    s2 = recurse(score_chart, curr_str, best+1, en)
+                    
+                    if mean:
+                        norm_score = best_score - (tot_score/(en-st))
+                    else:
+                        k_rand = randint(st, en-1)
+                        norm_score = best_score - score_chart[(st,k_rand)] - score_chart[(k_rand+1,en)]
+
+                norm_score = min(norm_score, self.hinge_const)
                 return s1 + s2 + norm_score
```

Comment:
Generally, try to avoid nesting large functions like this. Keep nested functions to a minimum, or avoid all together. This can stay here for now, but this makes the readability / usage of a function much less clear for reviewers. Shouldn't problem for our codebase if we do not keep adding on top ~ advice for the future!

Reply:
I based my code structure for this on Shikhar's initial attempt at the regularizer. I get your point though, this should definitely have been a separate function. Moved out into a separate method.